src/main/java/net/neoforged/neoforge/fluids/crafting/FluidIngredient.java

```diff
@@ -0,0 +1,292 @@
+     *           whether it is empty without resolving, you should override this method
+     *           with a custom implementation.
+     * @see #isEmpty()
+     */
+    public boolean hasNoFluids() {
+        return isEmpty() || getStacks().length == 0;
+    }
+
+    @Override
+    public abstract int hashCode();
+
+    @Override
+    public abstract boolean equals(Object obj);
+
+    // codecs
+    private static MapCodec<FluidIngredient> singleOrTagCodec() {
+        return NeoForgeExtraCodecs.xor(
+                SingleFluidIngredient.CODEC,
+                TagFluidIngredient.CODEC).xmap(either -> either.map(id -> id, id -> id), ingredient -> {
+                    if (ingredient instanceof SingleFluidIngredient fluid) {
+                        return Either.left(fluid);
+                    } else if (ingredient instanceof TagFluidIngredient tag) {
+                        return Either.right(tag);
+                    }
+                    throw new IllegalStateException("Basic fluid ingredient should be either a fluid or a tag!");
+                });
+    }
+
+    private static MapCodec<FluidIngredient> makeMapCodec() {
+        return NeoForgeExtraCodecs.<FluidIngredientType<?>, FluidIngredient, FluidIngredient>dispatchMapOrElse(
+                NeoForgeRegistries.FLUID_INGREDIENT_TYPES.byNameCodec(),
+                FluidIngredient::getType,
+                FluidIngredientType::codec,
+                FluidIngredient.SINGLE_OR_TAG_CODEC).xmap(either -> either.map(id -> id, id -> id), ingredient -> {
+                    // prefer serializing without a type field, if possible
+                    if (ingredient instanceof SingleFluidIngredient || ingredient instanceof TagFluidIngredient) {
+                        return Either.right(ingredient);
+                    }
+
+                    return Either.left(ingredient);
+                });
+    }
+
+    private static Codec<FluidIngredient> codec(boolean allowEmpty) {
+        var listCodec = Codec.lazyInitialized(() -> allowEmpty ? LIST_CODEC : LIST_CODEC_NON_EMPTY);
+        return Codec.either(listCodec, MAP_CODEC_CODEC)
+                // [{...}, {...}] is turned into a CompoundIngredient instance
+                .xmap(either -> either.map(CompoundFluidIngredient::of, i -> i),
+                        ingredient -> {
+                            // serialize CompoundIngredient instances as an array over their children
+                            if (ingredient instanceof CompoundFluidIngredient compound) {
+                                return Either.left(compound.children());
+                            } else if (ingredient.isEmpty()) {
+                                // serialize empty ingredients as []
+                                return Either.left(List.of());
+                            }
+
+                            return Either.right(ingredient);
+                        });
+    }
+
+    // empty
+    public static FluidIngredient empty() {
```

Comment:
I would move all these methods above the private static implementation ones for discoverability.